### ISSUE TITLE
Custom "Cache-Control" header seems to have stopped being set since May

### ISSUE THREAD

**Issue by ZYinMD** [zeit/now#7953]
In my web app, I aggressively cache some static assets in user's browsers by [setting headers](https://vercel.com/docs/project-configuration#project-configuration/headers) like `cache-control: public, max-age=2592000, stale-while-revalidate=31536000`. However, I noticed that in my recent deployments, this header was lost. It fell back to the default `cache-control: public, max-age=0, must-revalidate`.

I checked my git history and made sure I never changed my deploy configs, so I suspect something changed on vercel's side.

I always deploy from cli: `vercel --prod --local-config .vercel/vercel.json`. And my `vercel.json` is:
```json
{
  "headers": [
    {
      "source": "/assets/:any",
      "headers": [
        {
          "key": "Cache-Control",
          "value": "public, max-age=2592000, stale-while-revalidate=31536000"
        }
      ]
    }
  ]
}
```

I looked though all my deployments over time, and found the last deployment that had the correct header was on 4/24, all deployments before that were good. The first deployment that lost the header was on 5/16, all deployments after that didn't have the header. So, something changed between 4/24 and 5/16, maybe the vercel cli, or maybe the hosting server. 

I may have upgraded my vercel cli during that time, but I can't pinpoint the date.

**Comment by ZYinMD** [comment]
I found the answer - it has to do with the setting of `dist` folder.

Because I host a simple static website and I build locally, I don't need vercel to build for me in the cloud, so I set `build command` to `""`, then set both `Output Directory` and `Root Directory` to `"dist"`. It may look weird, but I think it makes sense, because vercel doesn't need to worry about anything outside of my locally bundled dist folder.

```
project-root
  |---.vercel/
  |---src/
  |---dist/
  |     |---index.html
  |     |---assets/
```
This has been working fine until May, then I believe there had been a regression - it still deploys ok, but my `headers` config doesn't work anymore.

Anyway, I found that if I change `Root Directory` to `"./"`, this issue is fixed.

**Comment by verbose117** [comment]
> In my web app, I aggressively cache some static assets in user's browsers by [setting headers](https://vercel.com/docs/project-configuration#project-configuration/headers) like `cache-control: public, max-age=2592000, stale-while-revalidate=31536000`. However, I noticed that in my recent deployments, this header was lost. It fell back to the default `cache-control: public, max-age=0, must-revalidate`.
> 
> I checked my git history and made sure I never changed my deploy configs, so I suspect something changed on vercel's side.
> 

When deployed, vercel is using the default cache control header that is clearly shown in the docs. You can instruct the Edge Network to use cache by using the directive `s-maxage`.

Read more about it in the docs: https://vercel.com/docs/concepts/edge-network/headers#cache-control-header.

**Comment by ZYinMD** [comment]
hi @haneenmahd , I was talking about caching in user's browser, not the edge network?

**Comment by verbose117** [comment]
Cache settings will be overwritten when in production to ensure static assets get cached effectively. So you can set it by using the function `getServerSideProps` and inside that function set the header for `Cache-Control`.

Like this
```js
export async function getServerSideProps({ req, res }) {
  res.setHeader(
    'Cache-Control',
    'public, s-maxage=2592000, stale-while-revalidate=31536000'
  )

  return {
    props: {},
  }
}
```

If you wanna read more about caching, see docs: https://nextjs.org/docs/going-to-production#caching

**Comment by ZYinMD** [comment]
@haneenmahd I think you and I are talking apples and oranges. I'm deploying a static site, not SSR, not nextjs, and not using functions. In the original issue I have already linked to the doc regarding http cache, and in my own reply I have identified the cause.

**Comment by haneenmahd** [comment]
Do you have like a repo to reproduce this issue

**Comment by verbose117** [comment]
Okay! so does this issue still need any fix?

**Comment by ZYinMD** [comment]
This can't be reproduced with a repo, because the `Root Directory` is configured in the dashboard, not in the repo. However, it's quite easy for you to reproduce with the configs mentioned above. 

As for whether it needs fix, it's up to the vercel team, but I think yes. As I said, 1. it used to work, and now it doesn't, so I think it's an regression. 2. According to the doc it should work, but it doesn't, and it's mysteriously related to the config of "Root Directory".

**Comment by verbose117** [comment]
Yes, as per the docs this should work but it doesn't. it's very confusing! but I will take a look at this 👀

**Comment by ZYinMD** [comment]
I noticed that this issues seems to have been fixed, the fix happened somewhere between 2022-07-22 and 2022-07-17. It seems to have happened on the hosting server and no upgrade to vercel CLI was needed.

**Comment by verbose117** [comment]
> I noticed that this issues seems to have been fixed, the fix happened somewhere between 2022-07-22 and 2022-07-17. It seems to have happened on the hosting server and no upgrade to vercel CLI was needed.

Okay!

**Comment by danieljaguiar** [comment]
I think this is not working again. Do we have any news here?

**Comment by wiiiimm** [comment]
> I think this is not working again. Do we have any news here?

@danieljaguiar can you describe your issue here and share your configurations with us?

**Comment by danieljaguiar** [comment]
@williamli 

Here is a simple repro: https://github.com/danieljaguiar/cache-control-test

I deployed it here: https://cache-control-test.vercel.app/

Three pages with 3 types of Cache-Control setting
- Cache Control set on getServerSideProps: https://cache-control-test.vercel.app/ssr-cache
- Cache Control set on next-config: https://cache-control-test.vercel.app/next-cache
- Cache Control set on vercel.json: https://cache-control-test.vercel.app/vercel-cache

I also set a "bonus" header called "Test-Type"

If you open devtools, the Cache Control is not set. Also, the "Test-Type" is set only on the getServerSideProps:

![image](https://user-images.githubusercontent.com/23394221/228547026-5b2305ea-cde2-46b8-8c18-4f08cc67b16c.png)

**Comment by maggch97** [comment]
There's a big issue when using vercel.json to set the Cache-Control header. Even the Cache-Control header is setted correctly, the age header will be larger than the max-age, then Chrome will ignore cache and request the resource every time.

**Comment by MarkoXR** [comment]
Any update on this, I am seeing the same behaviour?

**Comment by ustun-ed** [comment]
Seeing the same behavior in a custom SPA (not nextjs). `cache-control` in `vercel.json` is not being honored. I'm adding extra headers to see that the path matches correctly, it does, but cache-control is always the default one.

**Comment by ZYinMD** [comment]
It's working for me again. If you have this issue, try to make some minor modification to your `vercel.json` so it becomes a slightly different file, then deploy again. That seems to be what worked for me.

**Comment by antitoxic** [comment]
Two years later, I'm facing the same issue. Is there any solution for that? I tried the root directory change and tried to change the vercel.json file. No success. The cache headers are still the default "must-revalidate"
